### ISSUE TITLE
Trilinos nightly test failure cuda/10.1 on Power9+Volta70: TpetraCore_CrsMatrix_UnitTests_MPI_4

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4708]
Nightly builds of Trilinos with Kokkos' develop branch are failing in unit test `TpetraCore_CrsMatrix_UnitTests_MPI_4` in cuda/10.1 builds on Power9+Volta70 arch.

**Output snip**:
```
23:18:40 1685/2600 Test  #612: TpetraCore_CrsMatrix_UnitTests_MPI_4 ...........................................................***Failed  Required regular expression not found. Regex=[End Result: TEST PASSED
23:18:40 ]  1.87 sec
...
23:18:40 3. CrsMatrix_int_longlong_double_Kokkos_Compat_KokkosCudaWrapperNode_BadCalls_UnitTest ... 
23:18:40  Test that code {zero->apply(mv,mv);} throws std::runtime_error: passed
23:18:40  
23:18:40  Exception message for expected exception:
23:18:40  
23:18:40   /home/jenkins/weaver-new/workspace/KokkosEco_Trilinos_Weaver_CUDA101_opt/Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:5560:
23:18:40   
23:18:40   Throw number = 1
23:18:40   
23:18:40   Throw test that evaluated to true: ! isFillComplete ()
23:18:40   
23:18:40   Tpetra::CrsMatrix::apply: Cannot call apply() until fillComplete() has been called
...
23:18:40  zero->getFrobeniusNorm() = 0.00000000000000000e+00 == 0.00000000000000000e+00 : passed
23:18:40  Test that code {zero->apply(mv2,mv1);} throws std::runtime_error: failed (code did not throw an exception at all)
23:18:40  Test that code {zero->apply(mv2,mv3);} throws std::runtime_error: failed (code did not throw an exception at all)
...
```

**Reproducer** (weaver):
```
export ATDM_CONFIG_REGISTER_CUSTOM_CONFIG_DIR=${TRILINOS_DIR}/cmake/std/atdm/contributed/weaver
source ${TRILINOS_DIR}/cmake/std/atdm/load-env.sh weaver-cuda-10.1-opt
export ATDM_CONFIG_USE_NINJA=OFF

export KOKKOS_PATH=$HOME/kokkos
export OMPI_CXX=$KOKKOS_PATH/bin/nvcc_wrapper

rm -rf CMake*

# Configure
cmake \
 -G"Unix Makefiles" \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DCMAKE_INSTALL_PREFIX="${PWD}/install" \
 -DCMAKE_CXX_STANDARD="14" \
 -DCMAKE_CXX_FLAGS="-pedantic -Wall" \
 -DTrilinos_ENABLE_TESTS=OFF \
 -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
 -DTrilinos_ENABLE_Tpetra=ON \
 -DTpetra_ENABLE_TESTS=ON \
  -DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
  -DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR

make -j

ctest -R TpetraCore_CrsMatrix_UnitTests_MPI_4
```

I confirmed this fails with builds with Kokkos' develop branch; Trilinos' develop branch with unmodified Kokkos passed.

I'll need to bisect on Kokkos merges after Jan 10 to further triage, this crept in while the build was failing with #4677

**Comment by ajpowelsnl** [comment]
@ndellingwood -- could you please investigate a bit further to see how a Kokkos change caused this error?

**Comment by ndellingwood** [comment]
@ajpowelsnl yes in progress running bisection. Weaver is heavily congested so progress is a bit slow waiting on jobs to work their way through the queue

**Comment by ndellingwood** [comment]
Looks like this may have come from an update to KokkosKernels on 01/12/2022, kokkos/kokkos-kernels#1255

Here is a summary from investigating the breaking change:
- Final passing nightly of this Cuda build was job 399: https://jenkins-son.sandia.gov/view/Kokkos/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/399/
- Broken build at job 400, due to issue #4677; this build was down until the fix to Trilinos was merged for job 407
- The `TpetraCore_CrsMatrix_UnitTests_MPI_4` test failure was introduced by changes somewhere between nightly jobs 399 and 407

Bisection began with the following shas confirmed to have the `TpetraCore_CrsMatrix_UnitTests_MPI_4` test passing:
kokkos@ff4db5320bab77832c79d2c377cde1b49b25dcd9  (sha from job 399 mentioned above)
kokkos/kokkos-kernels@80b74ca726af801da34fb5539378977857d34914  (sha from job 399 mentioned above)
trilinos/Trilinos@ca382dbdb180a8a54c553d6677635091344184a4 (kokkos-promotion branch)

Keeping the kokkos and Trilinos shas fixed as above, the following shas correspond to when the breaking changes were introduced to kokkos-kernels
Passing sha at job 401: kokkos/kokkos-kernels@2a3e31c3911effd73c7c7b49b8e97da5ee10d4c1
Failing sha at job 402: kokkos/kokkos-kernels@ebec1a39eb0c783379f56d552335c18ebaf40209

Job 402 had two changes merged to kokkos-kernels, one updating the cm_test_all_sandia script (kokkos/kokkos-kernels@ebec1a39eb0c783379f56d552335c18ebaf40209) which is a testing script without any source code changes, and prior to that was PR kokkos/kokkos-kernels#1255

I'll migrate this issue and summary to kokkos-kernels, wanted to add update here for reference

@lucbv @brian-kelley @srajama1 help needed to resolve what changes from kokkos/kokkos-kernels#1255 are disrupting the Tpetra test. I skimmed over the PR, it is mostly changes related to bsr support (is that used by Tpetra yet?), but there are also some changes to the spmv files
